### ISSUE TITLE
Rebalance stuck, partitions rebalance not progressing in Kafka

### ISSUE THREAD

**Issue by amuraru** [linkedin/cruise-control#1167]
I am seeing an issue with the partition reassignments performed by CC: basically any rebalance that triggers more than one partition reassignment batch (tasks?) in kafka never completes.
It gets stuck in the second batch. with cc waiting for reassignment to complete and kafka controller doing nothing.
The `reassign_partition` zk node is being updated by CC but Kafka Controller is not triggering the rebalance.
The only workaround is to delete /controller in kafka

Working with kafka 2.4.1 server side and a build of cc that has the 2.4.1 client libs (master)

Looking at 
https://github.com/linkedin/cruise-control/blob/b386141146dfe4cc013b8233c19434e09b50027c/cruise-control/src/main/scala/com/linkedin/kafka/cruisecontrol/executor/ExecutorUtils.scala#L89 it seems that CC is not using `ReassignPartitionsCommand` but instead updating the zk node directly.
This apparently is not enough in newer version of Kafka.

**Comment by kyguy** [comment]
Hi @amuraru,

I was having trouble reproducing this issue and was hoping there might be more tips I could try!

Currently I am testing with Kafka 2.4.1 server side and CC with 2.4.1 client libraries. I am running a small 3 broker cluster with 66 partitions. I also set the `num.concurrent.partition.movements.per.broker` to 1 to ensure that the rebalance was batched.

Are there any other methods I could try to help trigger this failure?

**Comment by amuraru** [comment]
My reproducible environment is documented at https://github.com/amuraru/k8s-kafka-the-hard-way (CruiseControl integrated in Kafka Operator).
6 brokers cluster, 6 topics, ~430 replicas.
Any goal violation triggering a self healing balance reproduces this issue.
Reading through the CC code it seems to be that the  `executeReplicaReassignmentTasks` is actually updating in place the `/admin/reassign_partitions` zk node which is not something Kafka Controller supports.

**Comment by simonasr** [comment]
I guess we are having same issue. We launched cruise-control only yesterday.
First thing to try was to kick "Rebalance Broker Disks". It planned to move ~120 partitions, when done moving about ~30 of them ExecutorState changed to `STOPPING_EXECUTION`, status was showing there is 1 operation in progress, but there wasn't.
Couple hours later I restarted cruise control. Now ExecutorState shows `NO_TASK_IN_PROGRESS`. But now it complains:
```
ERROR: Error processing POST request '/rebalance' due to: 'com.linkedin.kafka.cruisecontrol.exception.OngoingExecutionException: There are ongoing intra-broker partition movements.'.
```

**Comment by amuraru** [comment]
@simonasr that's the behaviour we are observinbf as well - first reassignment passes and gets stuck in subsequent ones.
The only workaround is to `deleteall /admin/reassign_partitions` zk node in kafka and force a controller reelection by deleting the `/controller` zk node. That would help cc to make progress with the next batch of reassignments but eventually it will get stuck on next ones

**Comment by simonasr** [comment]
Thanks @amuraru ! Unfortunately this workaround didn't worked for me - `/admin/reassign_partition` was not existing.
Later I found, that `./kafka-log-dirs.sh` output shows one partition with attribute `"isFuture": true` and this was the reason preventing Kafka cc to start rebalancing. Moving out that partition to other broker unblocked Kafka CC 🤞

**Comment by simonasr** [comment]
This time it did more than 50% of the job :) and stuck again :(
```
[2020-04-17 15:12:32,622] INFO 65/114 (57.02%) intra-broker partition movements completed. 3199279/5666391 (56.46%) MB have been moved. (com.linkedin.kafka.cruisecontrol.executor.Executor)
``` 

Some logs on that moment, when CC decided to stop:
```
[2020-04-17 15:12:32,622] INFO 65/114 (57.02%) intra-broker partition movements completed. 3199279/5666391 (56.46%) MB have been moved. (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:12:32,622] INFO Completed tasks: [{EXE_ID: 384, INTRA_BROKER_REPLICA_ACTION(1002), {partition.name.edited-169, oldLeader: 1002, [1002, 1001, 1003] -> [1002, 1001, 1003]}, COMPLETED}]. (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:12:32,622] INFO Executor will execute 1 task(s) (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:13:23,935] INFO Skipping anomaly detection because the executor is in INTRA_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS state. (com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils)
[2020-04-17 15:13:42,727] WARN Encounter exception null when fetching logdir information for replica partition.name.edited-24-1003 (com.linkedin.kafka.cruisecontrol.executor.ExecutorAdminUtils)
[2020-04-17 15:13:45,897] INFO Post processing anomaly {Fixable broker failures detected: {Broker 1003 failed at 17/04/2020 15:13:45}} because executor is in INTRA_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS state. (com.linkedin.kafka.cruisecontrol.detector.AnomalyDetector)
[2020-04-17 15:13:46,303] WARN Killing execution for task {EXE_ID: 390, INTRA_BROKER_REPLICA_ACTION(1003), {partition.name.edited-24, oldLeader: 1001, [1001, 1002, 1003] -> [1001, 1002, 1003]}, DEAD} because the destination disk is down. (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:13:46,304] INFO 66/114 (57.89%) intra-broker partition movements completed. 3278434/5666391 (57.86%) MB have been moved. (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:13:46,304] INFO Completed tasks: [{EXE_ID: 390, INTRA_BROKER_REPLICA_ACTION(1003), {partition.name.edited-24, oldLeader: 1001, [1001, 1002, 1003] -> [1001, 1002, 1003]}, DEAD}].
[2020-04-17 15:13:46,304] INFO Stop the ongoing execution due to 1 dead or aborting tasks: [{EXE_ID: 390, INTRA_BROKER_REPLICA_ACTION(1003), {partition.name.edited-24, oldLeader: 1001, [1001, 1002, 1003] -> [1001, 1002, 1003]}, DEAD}]. (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:13:46,304] INFO Waiting for 7 tasks moving 545677 MB to finish (com.linkedin.kafka.cruisecontrol.executor.Executor)
[Dead/aborting: [390]] (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:14:06,339] WARN Encounter exception null when fetching logdir information for replica partition.name.edited-26-1003 (com.linkedin.kafka.cruisecontrol.executor.ExecutorAdminUtils)
[2020-04-17 15:14:06,340] INFO Completed tasks: [{EXE_ID: 401, INTRA_BROKER_REPLICA_ACTION(1003), {partition.name.edited-26, oldLeader: 1003, [1003, 1004, 1005] -> [1003, 1004, 1005]}, DEAD}].
[2020-04-17 15:14:06,340] INFO Waiting for 6 tasks moving 467176 MB to finish (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:14:06,340] WARN Killing execution for task {EXE_ID: 401, INTRA_BROKER_REPLICA_ACTION(1003), {partition.name.edited-26, oldLeader: 1003, [1003, 1004, 1005] -> [1003, 1004, 1005]}, DEAD} because the destination disk is down. (com.linkedin.kafka.cruisecontrol.executor.Executor)
[Dead/aborting: [401]] (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:14:16,473] INFO Skip sampling because the load monitor is in PAUSED state due to Paused-By-Cruise-Control-Before-Starting-Execution (Date: 2020-04-17_13:27:38 UTC).. (com.linkedin.kafka.cruisecontrol.monitor.task.SamplingTask)
[2020-04-17 15:15:22,026] INFO Skipping anomaly detection because the executor is in STOPPING_EXECUTION state. (com.linkedin.kafka.cruisecontrol.detector.AnomalyDetectorUtils)
[2020-04-17 15:19:36,848] INFO Completed tasks: [{EXE_ID: 370, INTRA_BROKER_REPLICA_ACTION(1004), {partition.name.edited-147, oldLeader: 1004, [1004, 1005, 1006] -> [1004, 1005, 1006]}, COMPLETED}]. (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:19:36,848] INFO Waiting for 5 tasks moving 389647 MB to finish (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:30:51,164] INFO Waiting for 4 tasks moving 309646 MB to finish (com.linkedin.kafka.cruisecontrol.executor.Executor)
[2020-04-17 15:30:51,164] INFO Completed tasks: [{EXE_ID: 387, INTRA_BROKER_REPLICA_ACTION(1002), {partition.name.edited-4, oldLeader: 1005, [1005, 1001, 1002] -> [1005, 1001, 1002]}, COMPLETED}]. (com.linkedin.kafka.cruisecontrol.executor.Executor)
```

Now `kafka-log-dirs.sh` output includes six partitions with `"isFuture": true`.

**Comment by amuraru** [comment]
One user pointed to https://issues.apache.org/jira/browse/KAFKA-9478 on gitter as the possible cause for this

**Comment by amuraru** [comment]
That being said I am hopeful a committer here reviews #1168 and cc starts using the public reassignment "api" and messing with the underlying zk node directly

**Comment by efeg** [comment]
Hi @amuraru 
Thanks for creating this issue, linking the relevant Kafka ticket, and submitting #1168 with a proposed mitigation.

Cruise Control (CC) directly uses the underlying reassignment `zNode` rather than the public API because, today Kafka lacks the support for CC to adjust the reassignment concurrency dynamically -- i.e. while there are ongoing replica movements. The existing `reassignPartitionsCommand` does not allow dynamically adding more replica moves while there are ongoing replica moves. Note that due to a behavior change in Kafka controller since `1.1.0+`, this concurrency adjustment works as expected only if the workaround described in https://github.com/linkedin/cruise-control/issues/496 (see [the patch](https://github.com/linkedin/kafka/commit/8f254b952ca29ad1d406b1a075ff387ba6686389) adopted by LinkedIn Kafka) is used.

Looks like https://issues.apache.org/jira/browse/KAFKA-9478 might indeed have introduced another glitch in replica reassignment -- I will look into it more closely.

Finally, [KIP-455](https://cwiki.apache.org/confluence/display/KAFKA/KIP-455%3A+Create+an+Administrative+API+for+Replica+Reassignment) was initially intended to be a part of the `2.4.0` release, but the KIP page indicates plans to support it in `2.6.0`. Once it is supported, CC will be able to use its public API for replica moves and cancelling ongoing moves. 

@stanislavkozlovski I see that some controller changes regarding `KIP-455` were included in `2.4.0`. Would you be able to comment on the current usability of these APIs? Can CC adopt existing KIP-455 changes in `2.4.0`?

**Comment by efeg** [comment]
This issue has intended to be addressed in `migrate_to_kafka_2_4` branch starting with the version https://github.com/linkedin/cruise-control/releases/tag/2.4.1

-- Closing this issue. Please let us know if you see any issues on `Kafka 2.4+` using `2.4.*` releases of CC.

**Comment by isugimpy** [comment]
A little late on this one since it's closed, but running CC 2.4.8 on Kafka 2.4.0, I'm still experiencing this behavior.  Would it be possible to reopen this issue, or should I open a new one?

**Comment by ccrootdeeplay** [comment]
So after some yaers )) i'll have a similar problem in kafka 3.2.0 with cruise-control-core-2.5.95.
